src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php

```diff
@@ -276,7 +276,7 @@ private function hasUserChanged(UserInterface $user): bool
             $userRoles[] = 'ROLE_PREVIOUS_ADMIN';
         }
 
-        if (\count($userRoles) !== \count($this->getRoleNames()) || \count($userRoles) !== \count(array_intersect($userRoles, $this->getRoleNames()))) {
+        if (\count($this->getRoleNames()) !== \count(array_intersect($this->getRoleNames(), $userRoles))) {
```

Comment:
Can you please update also https://github.com/symfony/symfony/blob/4f4c30d59ed429bc08a34f330c45de2b04059d0a/src/Symfony/Component/Security/Core/User/User.php#L148

Reply:
Done https://github.com/gorshkov-ag/symfony/commit/429dee7efb5d018a67d0654bfbc956cb08d1b86c

Reply:
Now is it problem in User equals test: User without role now is equivalent with user with role "ROLE", but opposite is not true.
https://github.com/gorshkov-ag/symfony/blob/9de09f0cfbfed2af1f7ae0a63597171615a51ac3/src/Symfony/Component/Security/Core/Tests/User/UserTest.php#L110-L120
120 line case.